\documentclass{article}

% Set page size and margins
\usepackage[a4paper,top=2cm,bottom=2cm,left=2.5cm,right=2.5cm,marginparwidth=1.75cm]{geometry}

% Useful packages

\usepackage{graphicx}
%\usepackage[draft]{graphicx}

\usepackage{amsmath}
\usepackage{wrapfig}
\usepackage{subcaption}
\usepackage{caption}
\usepackage{float}
\usepackage{multicol}
\usepackage[colorlinks=true, allcolors=blue]{hyperref}
\usepackage{lscape} 

\renewcommand\arraystretch{1.25}

\usepackage{listings}
\usepackage{color}

\definecolor{dkgreen}{rgb}{0,0.6,0}
\definecolor{gray}{rgb}{0.5,0.5,0.5}
\definecolor{mauve}{rgb}{0.58,0,0.82}

\lstset{frame=tb,
  language=R,
  aboveskip=3mm,
  belowskip=3mm,
  showstringspaces=false,
  columns=flexible,
  basicstyle={\small\ttfamily},
  numbers=none,
  numberstyle=\tiny\color{gray},
  keywordstyle=\color{blue},
  commentstyle=\color{dkgreen},
  stringstyle=\color{mauve},
  breaklines=true,
  breakatwhitespace=true,
  tabsize=3
}

%\numberwithout{figure}{section}
\counterwithout*{figure}{section} % this is the only way i got the figure number and in-text reference to match, no idea what kind of black magic is happening here though


\title{How many genes are in a gene family? Uniform annotation methods for comparative genomics}
\author{Milena Trabert, Jesper Bomann, Elina Immonen}
\date{May 2025}

\begin{document}
\pagenumbering{gobble}



\begin{figure}[H]

    \begin{subfigure}[t]{\textwidth}
    %\makebox[0pt][l]{\hspace{200pt}\textbf{A}}
    \begin{picture}(0,0)
   	  		\put(30,-30){\textbf{A}} % adjust the x and y values here
		\end{picture}
        \begin{center}
        \includegraphics[width=\linewidth]{../data/fastX_ortholog_ident/fastX_permutation_bruchini.png}
        \end{center}        
        %\caption{ } \label{OG_stats:num_genes}
    \end{subfigure}

    \hfill

    \begin{subfigure}[t]{0.49\linewidth}
	    %\raggedleft
	    %\makebox[0pt][l]{\hspace{30pt}\textbf{C}}
	    \begin{picture}(0,0)
   	  		\put(30,-5){\textbf{B}} % adjust the x and y values here
		\end{picture}
	    \begin{center}
        \includegraphics[width=0.95\linewidth]{../data/fastX_ortholog_ident/fastX_permutation_coccinella.png}
        \end{center}        
        %\caption{ } \label{OG_stats:GF_vs_GS}
    \end{subfigure}
    \begin{subfigure}[t]{0.49\linewidth}
    %\makebox[0pt][l]{\hspace{30pt}\textbf{D}}
    \begin{picture}(0,0)
   	  		\put(30,-5){\textbf{C}} % adjust the x and y values here
		\end{picture}
	    \begin{center}
        \includegraphics[width=0.95\linewidth]{../data/fastX_ortholog_ident/fastX_permutation_tribolium.png}
        \end{center}        
        %\caption{ } \label{OG_stats:GF_vs_rep}
    \end{subfigure}
    
    
    \caption{\textbf{d$_N$/d$_S$ ratio and permutation tests for significance.} Pairwise comparisons between members of three species groups with violin plots of d$_N$/d$_S$ values for X-linked and autosomal 1-to-1 orthologs, and permutation tests (10000 permutations) to assess significance. All pairwise comparisons show significantly lower d$_N$/d$_S$ values on X-linked orthologs. The within-family comparisons are performed for three species groups: \textit{Bruchini} (\textbf{A}), \textit{Coccinella} (\textbf{B}) and \textit{Tribolium} (\textbf{C}).
    }\label{dNdS_ratios}
    
\end{figure}




\begin{figure}[H]

    \begin{subfigure}[t]{\textwidth}
    %\makebox[0pt][l]{\hspace{200pt}\textbf{A}}
    \begin{picture}(0,0)
   	  		\put(30,-30){\textbf{A}} % adjust the x and y values here
		\end{picture}
        \begin{center}
        \includegraphics[width=\linewidth]{../data/fastX_ortholog_ident/dS_vs_dN_scatterplot_bruchini.png}
        \end{center}        
        %\caption{ } \label{OG_stats:num_genes}
    \end{subfigure}

    \hfill

    \begin{subfigure}[t]{0.49\linewidth}
	    %\raggedleft
	    %\makebox[0pt][l]{\hspace{30pt}\textbf{C}}
	    \begin{picture}(0,0)
   	  		\put(30,-5){\textbf{B}} % adjust the x and y values here
		\end{picture}
	    \begin{center}
        \includegraphics[width=0.95\linewidth]{../data/fastX_ortholog_ident/dS_vs_dN_scatterplot_coccinella.png}
        \end{center}        
        %\caption{ } \label{OG_stats:GF_vs_GS}
    \end{subfigure}
    \begin{subfigure}[t]{0.49\linewidth}
    %\makebox[0pt][l]{\hspace{30pt}\textbf{D}}
    \begin{picture}(0,0)
   	  		\put(30,-5){\textbf{C}} % adjust the x and y values here
		\end{picture}
	    \begin{center}
        \includegraphics[width=0.95\linewidth]{../data/fastX_ortholog_ident/dS_vs_dN_scatterplot_tribolium.png}
        \end{center}        
        %\caption{ } \label{OG_stats:GF_vs_rep}
    \end{subfigure}
    
    
    \caption{\textbf{d$_S$ vs. d$_N$ scatterplots and violin plots of d$_S$ values from X-linked and autosomal orthologs.} Pairwise comparisons between members of three species groups with violin plots of d$_S$ values for X-linked and autosomal 1-to-1 orthologs, and scatterplots of d$_S$ vs. d$_N$. d$_S$ is lower for X-linked orthologs in all comparisons except \textit{Tribolium}, where it is almost equal. The within-family comparisons are performed for three species groups: \textit{Bruchini} (\textbf{A}), \textit{Coccinella} (\textbf{B}) and \textit{Tribolium} (\textbf{C}).
    }\label{dNdS_ratios}
    
\end{figure}






\end{document}