\documentclass{article}

% Set page size and margins
\usepackage[a4paper,top=2cm,bottom=2cm,left=2.5cm,right=2.5cm,marginparwidth=1.75cm]{geometry}

% Useful packages

\usepackage{graphicx}
%\usepackage[draft]{graphicx}

\usepackage{amsmath}
\usepackage{wrapfig}
\usepackage{subcaption}
\usepackage{caption}
\usepackage{float}
\usepackage{multicol}
\usepackage[colorlinks=true, allcolors=blue]{hyperref}
\usepackage{lscape} 

\renewcommand\arraystretch{1.25}

\usepackage{listings}
\usepackage{color}
\usepackage[percent]{overpic}

\definecolor{dkgreen}{rgb}{0,0.6,0}
\definecolor{gray}{rgb}{0.5,0.5,0.5}
\definecolor{mauve}{rgb}{0.58,0,0.82}

\lstset{frame=tb,
  language=R,
  aboveskip=3mm,
  belowskip=3mm,
  showstringspaces=false,
  columns=flexible,
  basicstyle={\small\ttfamily},
  numbers=none,
  numberstyle=\tiny\color{gray},
  keywordstyle=\color{blue},
  commentstyle=\color{dkgreen},
  stringstyle=\color{mauve},
  breaklines=true,
  breakatwhitespace=true,
  tabsize=3
}

%\numberwithout{figure}{section}
\counterwithout*{figure}{section} % this is the only way i got the figure number and in-text reference to match, no idea what kind of black magic is happening here though


\title{How many genes are in a gene family? Uniform annotation methods for comparative genomics}
\author{Milena Trabert, Jesper Bomann, Elina Immonen}
\date{May 2025}

\begin{document}
\pagenumbering{gobble}

% pdflatex -pdf composite_figures.tex




\begin{figure}[H]

    \begin{subfigure}[t]{\textwidth}
    %\makebox[0pt][l]{\hspace{200pt}\textbf{A}}
    \begin{picture}(0,0)
   	  		\put(30,-30){\textbf{A}} % adjust the x and y values here
		\end{picture}
        \begin{center}
        \includegraphics[width=\linewidth]{../data/fastX_ortholog_ident/pos_sel_dNdS_summary_plot_bruchini.png}
        \end{center}        
        %\caption{ } \label{OG_stats:num_genes}
    \end{subfigure}

    \hfill

    \begin{subfigure}[t]{0.49\linewidth}
	    %\raggedleft
	    %\makebox[0pt][l]{\hspace{30pt}\textbf{C}}
	    \begin{picture}(0,0)
   	  		\put(30,-5){\textbf{B}} % adjust the x and y values here
		\end{picture}
	    \begin{center}
        \includegraphics[width=0.95\linewidth]{../data/fastX_ortholog_ident/pos_sel_dNdS_summarylot_coccinella.png}
        \end{center}        
        %\caption{ } \label{OG_stats:GF_vs_GS}
    \end{subfigure}
    \begin{subfigure}[t]{0.49\linewidth}
    %\makebox[0pt][l]{\hspace{30pt}\textbf{D}}
    \begin{picture}(0,0)
   	  		\put(30,-5){\textbf{C}} % adjust the x and y values here
		\end{picture}
	    \begin{center}
        \includegraphics[width=0.95\linewidth]{../data/fastX_ortholog_ident/pos_sel_dNdS_summaryplot_tribolium.png}
        \end{center}        
        %\caption{ } \label{OG_stats:GF_vs_rep}
    \end{subfigure}
    
    
    \caption{\textbf{d$_N$/d$_S$ and abundance of positively selected genes.} 
    \textit{Bottom left}: Pairwise comparisons between members of three species groups with violin plots of d$_N$/d$_S$ values for X-linked and 
    autosomal 1-to-1 orthologs, and scatterplots of d$_S$ vs. d$_N$. Permutation tests (n=10000) show that d$_N$/d$_S$ is 
    significantly lower for X-linked orthologs in all comparisons within \textit{Bruchini} (\textbf{A}) and 
    \textit{Coccinella} (\textbf{B}), but not \textit{Tribolium} (\textbf{C}). \textit{Top right}: The bar plot show the percentage of genes that 
    contain codons under positive selection according to paml site-model analysis (likelihood ratio test comparison between 
    paml site models M1a and M2a). \textit{Bruchini} (\textbf{A}), show an enrichment of positively selected X-linked orthologs in all pairwise comparisons, 
    while \textit{Coccinella} (\textbf{B}) and \textit{Tribolium} (\textbf{C}) show no significant difference.
    }\label{dS_plots}
    
\end{figure}






\begin{figure}[H]
    

    \begin{subfigure}[t]{0.49\linewidth}
	    %\raggedleft
	    %\makebox[0pt][l]{\hspace{30pt}\textbf{C}}
	    \begin{picture}(0,0)
   	  		\put(30,-5){\textbf{B}} % adjust the x and y values here
		\end{picture}
	    \begin{center}
        \includegraphics[width=0.95\linewidth]{../data/DE_analysis/all_sex_bias_proportion.png}
        \end{center}        
        %\caption{ } \label{OG_stats:GF_vs_GS}
    \end{subfigure}
    \begin{subfigure}[t]{0.49\linewidth}
    %\makebox[0pt][l]{\hspace{30pt}\textbf{D}}
    \begin{picture}(0,0)
   	  		\put(30,-5){\textbf{C}} % adjust the x and y values here
		\end{picture}
	    \begin{center}
        \includegraphics[width=0.95\linewidth]{../data/DE_analysis/dNdS_A_obtectus_vs_sig_logFC.png}
        \end{center}        
        %\caption{ } \label{OG_stats:GF_vs_rep}
    \end{subfigure}

    \hfill

    \begin{subfigure}[t]{0.49\linewidth}
	    %\raggedleft
	    %\makebox[0pt][l]{\hspace{30pt}\textbf{C}}
	    \begin{picture}(0,0)
   	  		\put(30,-5){\textbf{D}} % adjust the x and y values here
		\end{picture}
	    \begin{center}
        \includegraphics[width=0.95\linewidth]{../data/DE_analysis/dNdS_B_siliquastri_vs_sig_logFC.png}
        \end{center}        
        %\caption{ } \label{OG_stats:GF_vs_GS}
    \end{subfigure}
    \begin{subfigure}[t]{0.49\linewidth}
    %\makebox[0pt][l]{\hspace{30pt}\textbf{D}}
    \begin{picture}(0,0)
   	  		\put(30,-5){\textbf{E}} % adjust the x and y values here
		\end{picture}
	    \begin{center}
        \includegraphics[width=0.95\linewidth]{../data/DE_analysis/dNdS_C_chinensis_vs_sig_logFC.png}
        \end{center}        
        %\caption{ } \label{OG_stats:GF_vs_rep}
    \end{subfigure}
    
    \caption{\textbf{Differential expression analysis.} 
    We are utilizing the log2 fold change (og2LFC) information from genes in \textit{C. maculatus} and combining it with 
    d$_N$/d$_S$ between these genes and their 1-to-1 orthologs in other bruchids. 
    The log2FC contrast is always female-male,therefore positive values indicate female-bias and negative values 
    indicate male-bias.
    \textbf{A}: log2FC of all genes on X-linked contigs in somatic (head+thorax) and reproductive (abdominal) tissues. 
    Positive log2FC indicates female bias. Vertical lines are the median log2FC for each tissue, and the shaded area is the
    standard error of that mean. 
    \textbf{B}: number of sex biased genes in abdominal (reproductive) tissues and head and thorax (somatic) tissues. 
    \textbf{C-E}: Scatterplot of significantly differentially expressed genes showing d$_N$/d$_S$ of 
    \textit{A. obtectus} (\textbf{C}), \textit{B. siliquastri} (\textbf{D}) and \textit{C. chinensis} (\textbf{E}) respectively.
    }\label{DE_analysis}
    
\end{figure}






\end{document}